### PR TITLE
YARN-6151. FS Preemption doesn't filter out queues which cannot be pr…

### PR DESCRIPTION
**Pull request by flyrain**
…eempted.

### CODE REVIEW

File: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSQueue.java [hunk 1]
```diff
@@ -236,6 +236,29 @@ public void setFairSharePreemptionThreshold(float fairSharePreemptionThreshold)
   }
 
   /**
+   * Recursively check if the queue can be preempted based on whether the
+   * resource usage is greater than fair share.
+   *
+   * @return true if the queue can be preempted
+   */
+  public boolean canBePreempted() {
+    assert parent != null;
```
Comment (kambatla) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSQueue.java]: Why is this necessary?

File: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/test/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/TestFairScheduler.java [hunk 2]
```diff
@@ -2036,10 +2036,10 @@ public void testPreemptionIsNotDelayedToNextRound() throws Exception {
...
     assertEquals(3277, toPreempt.getMemorySize());
 
-    // verify if the 3 containers required by queueA2 are preempted in the same
+    // verify if the 4 containers required by queueA2 are preempted in the same
     // round
     scheduler.preemptResources(toPreempt);
-    assertEquals(3, scheduler.getSchedulerApp(app1).getPreemptionContainers()
+    assertEquals(4, scheduler.getSchedulerApp(app1).getPreemptionContainers()
         .size());
   }
```
Comment (kambatla) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/test/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/TestFairScheduler.java]: Can we add a new test that verifies the exact scenario in the JIRA description?

File: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSQueue.java [hunk 3]
```diff
@@ -236,6 +236,29 @@ public void setFairSharePreemptionThreshold(float fairSharePreemptionThreshold)
   }
 
   /**
+   * Recursively check if the queue can be preempted based on whether the
+   * resource usage is greater than fair share.
+   *
+   * @return true if the queue can be preempted
+   */
+  public boolean canBePreempted() {
```
Comment (kambatla) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSQueue.java]: Should the check of the allowPreemptionFrom flag also be part of this method?

Comment (flyrain) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSQueue.java]: It should be, but allowPreemptionFrom is introduced after 2.8.x.

Comment (kambatla) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSQueue.java]: Aah, I keep forgetting branch-2.8 was cut years ago. :(

File: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/test/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/TestFairScheduler.java [hunk 4]
```diff
@@ -2043,6 +2043,78 @@ public void testPreemptionIsNotDelayedToNextRound() throws Exception {
...
 
+  @Test
+  public void testPreemptionFilterOutNonPreemptableQueues() throws Exception {
```
Comment (kambatla) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/test/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/TestFairScheduler.java]: Can we add this test to TestFairSchedulerPreemption instead?

File: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSQueue.java [hunk 5]
```diff
@@ -242,7 +244,9 @@ public void setFairSharePreemptionThreshold(float fairSharePreemptionThreshold)
    * @return true if the queue can be preempted
    */
   public boolean canBePreempted() {
-    assert parent != null;
+    Preconditions.checkNotNull(parent, "Parent queue can't be null since"
```
Comment (kambatla) [inline hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSQueue.java]: Maybe, we could make this message more clear. "Parent queue is null. Looks like we are checking if root can be preempted."

Alternatively, can we make the if check (parent != null && ...)? That way, else would capture the null case and things should work fine?

**Comment by flyrain** [comment]
Committed.
